src/tensor/private/arraymancer_exceptions.nim

```diff
@@ -0,0 +1,100 @@
+# Copyright 2017 the Arraymancer contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+#[
+
+########### Arraymancer Exceptions ############
+
+Author: Kaan Eraslan
+
+Purpose: Implement exceptions that are more specific to tensors and
+their usecases.
+This file also contains the hierarchy of arraymancer specific exceptions.
+
+
+Exception Hierarchy
+---------------------
+
+1. IndexError
+
+    1. SliceError
+
+        1. SliceStepError
+
+
+2. ValueError
+
+    1. IncompatibleShapeError
+
+        1. ElementWiseShapeError
+
+    2. IncompatibleSizeError
+
+        1. IncompatibleSizeReshapeError
+
+    3. ArgRankError
+
+    4. AxisRankError
+
+3. TypeError
+
+    1. TensorLayoutTypeError
+
+    2. IncompatibleRankError
+
```

Comment:
I prefer if both inherits from `ValueError`.
Also can you rename `TensorLayoutTypeError` to `TensorLayoutError`